kep/tbd-peaks-power-aware-scheduling/README.md

```diff
@@ -0,0 +1,326 @@
+
+### Notes/Constraints/Caveats
+
+- The [Min, Max] score range for the plugin is user-configurable, defaulting to [0, 100].
+- Additionally, the plugin normalizes the generated scores within the supplied [Min, Max] range.
+- There isn't a singular model that comprehensively depicts the correlation between utilization and power consumption for every cluster node.
+- At any given time, the current utilization of a node can be identified through metrics data.
+- Determining the current power consumption can be achieved either through live metrics or by inference from the model that describes the relationship between node utilization and power consumption.
+- Retraining the model representing this relationship may be necessary to better align with the workloads running on the node.
+
+### Risks and Mitigations
+
+- The benchmark test below captures the overhead resulting from model inferencing latency, which appears to be negligible.
+- The accuracy of the model that captures the relationship between utilization and power relies heavily on the quality of the metrics used for its training. The current implementation supports integration with various user-selected metrics providers, such as node-metrics and Kepler
+- PEAKS plugin requires the power model parameters for each cluster node configured with it. However, the power model (utilization vs. power consumption) parameters are subjected to change (e.g., increase in the number of CPUs on a node results in a change in the power model) over time. To address this, `configmap` is used as a mechanism to pass the updated power model parameters of any cluster node seamlessly.
+
+## Design Details
+
+### PEAKS workflow
+![PEAKS workflow](./figs/PEAKS_workflow.png)
+
+Here is a brief discription of some of the steps in the above workflow:
+- Create a power model for each cluster node: If a power model suitable for the workload to be scheduled already exists, then the same can be used to avoid creating a new power model. Thus it is an optional step.
+  - PEAKS plugin allows the cluster owners to bring their own power models that best represent both the node behavior and the workload characteristics.
+  - Training the power model is outside the scope of PEAKS plugin.
+  - PEAKS plugin only does power model inferencing.
+  - Here is an example that models the "CPU utilization vs. Power" relationship for a two node cluster.
+    - This model changes corresponding to a change either in the node configuration or in the workload characteristics.![Power-models-of-nodes](./figs/PowerModelsOfNodes.png)
```

Comment:
Can I interpret the model as a service that accept a pod as input, and returns a node's estimated power utilization (once the pod is scheduled) as output? If so, where is this model service hosted?

Reply:
Hi @Huang-Wei , the model is not a service. We have added below details explaining what a power model means, how to learn the model parameters and where the model inference happens.

=====
The power model typically is a mathematical expression (e.g., NodePower = K0 + K1 * e^(K2 * x), where x is node utilisation and each K is a constant).

- Training the power model implies learning the values of the parameters of the corresponding mathematical expression (e.g., learning the values of K0, K1 and K2 above).
- Inferencing the power model implies computing the value of the expression representing the node power by supplying the value of the node utilization (e.g., evaluating the function NodePower by supplying the value of the parameter x).
- Kubernetes API object ConfigMap can be used as a mechanism to pass the power model parameters of any cluster node seamlessly

Power model inferencing is part of the function that implements the [ScorePlugin](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/interface.go#L612) interface function Score of the PEAKS plugin.

=====

Hope these details address your comment.

Reply:
Gotcha, thanks for the explanation.

Reply:
I will check later your impl. to see if the ConfigMap can be simplified as plugin arguments.

Reply:
Sure, @Huang-Wei ! Thank you.